### PR TITLE
test(StatusInput): add regression unit test for StatusInput

### PR DESCRIPTION
**Pull request by stefandunca**
Adds regression test for #584

Add test helpers to monitor QML output using a declarative approach.
Improves the QML tests structure.

### CODE REVIEW

**Comment by PascalPrecht** [comment]
I'm a little late to the party, but can you tell me how I can verify that these tests work? We should update the README with instructions as well (can be part of another PR)

File: tests/TestControls/tst_test-statusinput.qml [hunk 1]
```diff
@@ -32,6 +34,13 @@ Item {
 
         when: windowShown
 
+        //> Simulate key and wait for side effects
+        function pressKeyAndWait(key, time) {
```
Comment (alaibe) [inline tests/TestControls/tst_test-statusinput.qml:38]: will this be usefull for many test? I think we just duplicated this function, maybe we can extract this?

**Review by alaibe — APPROVED** [review]

**Comment by stefandunca** [comment]
> I'm a little late to the party, but can you tell me how I can verify that these tests work? We should update the README with instructions as well (can be part of another PR)

@PascalPrecht I added this readme file: https://github.com/status-im/StatusQ/pull/590/files#diff-4cfdda6d53eac90934bf621814cd6f2166f91078fac58c2f56f53224a3c5abaa

Let me know if clear enough

File: tests/TestControls/tst_test-statusinput.qml (hunk 1 above)
Comment (stefandunca) [reply tests/TestControls/tst_test-statusinput.qml]: Done. Thanks

File: tests/TestControls/TestHelpers/MonitorQtOutput.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    Q_INVOKABLE QString qtOuput();
+
+signals:
+
+private:
+    static void qtMessageOutput(QtMsgType type, const QMessageLogContext &context, const QString &msg);
+    // Use it to keep track of qInstallMessageHandler call
+    static std::weak_ptr<QString> _qtMessageOutputForSharing;
+    static std::mutex _mutex;
+    std::shared_ptr<QString> _thisMessageOutput;
```
Comment (osmaczko) [inline tests/TestControls/TestHelpers/MonitorQtOutput.h:29]: We don't have coding guidelines in place, but for naming conventions, I think we should follow: https://github.com/status-im/status-desktop/tree/fix/cpp-structure. Or we can agree on using "_" prefix for private members. All I want to say is that we should stay consistent across all cpp source code across Status projects.

File: tests/TestControls/tst_test-StatusBaseInput.qml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        //
+        // Tests
+
+        function test_initial_empty_is_valid() {
+            mouseClick(statusInput)
+            // Do some editing
+            TestUtils.pressKeyAndWait(testCase, statusInput, Qt.Key_B)
+            TestUtils.pressKeyAndWait(testCase, statusInput, Qt.Key_Left)
+            TestUtils.pressKeyAndWait(testCase, statusInput, Qt.Key_A)
+            verify(qtOuput.qtOuput().length === 0, "No output expected. Check qtOuput.qtOuput()")
```
Comment (osmaczko) [inline tests/TestControls/tst_test-StatusBaseInput.qml:45]: Maybe we could print here the unexpected output directly?

**Review by osmaczko — APPROVED** [review]
That's great! Thanks to that we'll be able to catch any warnings for StatusQ reusable components in an automated way.

In the future we should:
- integrate it with CI
- add 'warnings free' tests for all StatusQ components

File: tests/TestControls/TestHelpers/MonitorQtOutput.h (hunk 2 above)
Comment (stefandunca) [reply tests/TestControls/TestHelpers/MonitorQtOutput.h:29]: Thanks for the heads up. I see Hungarian notation and no decoration approach in the same file: https://github.com/status-im/status-desktop/blob/185457d94fb3f4718eed3df12d3af61d739aaaa8/src-cpp-structure/projects/App/Core/StatusSyntaxHighlighter.h ; Wondering which is preferred, I would go fo no decoration.

@iurimatias codding convention looks like a good topic to address in one of the next C++ architecture meetings

File: tests/TestControls/tst_test-StatusBaseInput.qml (hunk 3 above)
Comment (stefandunca) [reply tests/TestControls/tst_test-StatusBaseInput.qml:45]: Will have to check.

File: tests/TestControls/tst_test-statusinput.qml [hunk 4]
```diff
@@ -4,6 +4,8 @@ import QtTest 1.0
 import StatusQ.Controls 0.1
```
[the remark this answers is not in the document]
Comment (stefandunca) [reply tests/TestControls/tst_test-statusinput.qml:4]: Will fix

**Pull request merged**